sample_annotator/ontology/triad_normalizer.py

```diff
@@ -0,0 +1,151 @@
+"""
+examine strings from biosample metadata and return matching MIxS traid terms if possible
+
+MIxS trad columns = env_broad_scale, env_local_scale, env_medium
+Might want to look at the harmonized_table.db created by https://github.com/INCATools/biosample-analysis
+path = XXX
+query = XXX
+
+Basic methods are provided fro extracting OBO foundry term IDs from strings,
+including obtaining label/term reference data
+Does not yet obtain synonyms etc.
+alternatives: deeper parse of JSON
+direct rdftab, semantic sql... setup
+ols, bioportal term lookup or search... slower
+"""
+import json
+import re
+from typing import Optional
+
+import pandas as pd
+import requests
+from src.curieutil import CurieUtil
+from strsimpy.cosine import Cosine
+
+from sample_annotator.ontology.package_checklist_normalizer import standard_tidy
+
+prefix_commons_context_url = 'https://raw.githubusercontent.com/prefixcommons/biocontext/master/registry' \
+                             '/obo_context.jsonld'
+envo_json_url = "http://purl.obolibrary.org/obo/envo.json"
+cosine_ngram_size = 2
+qualified_stringdist = Cosine(cosine_ngram_size)
+
+
+# put some caching mechanism in place
+# what to do with NULLs and empty strings?
+# also see notes in sample_annotator/ontology/package_checklist_normalizer.py
+
+def get_web_json(url: str) -> dict:
+    r = requests.get(url)
+    t = r.text
+    json_dict = json.loads(t)
+    return json_dict
+
+
+prefix_commons_context_res = requests.get(prefix_commons_context_url)
+prefix_commons_context_mapping = CurieUtil.parseContext(prefix_commons_context_res.json())
+curie_mapper = CurieUtil(prefix_commons_context_mapping)
+
+envo_dict = get_web_json(envo_json_url)
+envo_label_frame = pd.DataFrame(envo_dict['graphs'][0]['nodes'])
+envo_label_frame = envo_label_frame[['id', 'lbl']]
+# convert IRIs <-> CURIEs
+envo_label_frame['curie'] = envo_label_frame['id'].apply(curie_mapper.getCurie)
+envo_label_frame['tidied'] = envo_label_frame['lbl'].apply(standard_tidy)
+
+
+def get_bio_registry_content(url="https://raw.githubusercontent.com/bioregistry/bioregistry/main/src/bioregistry/data"
+                                 "/bioregistry.json") -> dict:
+    br_dict = get_web_json(url)
+    return br_dict
+
+
+# this should be global ie not re-downloaded and recreated for each normalization
+bio_registry_content = get_bio_registry_content()
```

Comment:
not totally clear why we need this?

Also I prefer not invoking a service call when loading a module. If it's necessary I would invoke the call instead on creation of an instance of a normalizer object

Reply:
Hi, @cmungall . Thanks but I don't understand your second point. I guess we can discuss this this afternoon.

Reply:
I'm using bioregistry to determine the pattern for term IDs, so that they can be extracted from submitted MIxS terms and handled differently from the label-like portions. Previously I had been processing all IDs from a given ontology with [TDDA rexpy](https://tdda.readthedocs.io/en/v1.0.30/rexpy.html) in order to determine the ID pattern.

I thought this was an improvement in terms of performance and memory usage.

Reply:
use bioregistry python library
if I was going to use a service, add arg for passing bioregistry object